### PR TITLE
Support upgrading non eksctl clusters

### PR DESCRIPTION
**Pull request by aclevername**
### Description
`eksctl upgrade cluster` can update clusters that aren't created via eksctl (e.g. created manually in the console)

The current upgrade flow for eksctl created cluster also checks that all of the [nodegroups are compatible](https://github.com/weaveworks/eksctl/blob/master/pkg/eks/compatibility.go#L45-L101). I have not added the same compatibility tests to the upgrade path for non-eksctl support as you can't currently create nodegroups for these clusters, so it can't contain any incompatible nodes. This is something we should be aware of in the future if we add support for creating nodegroups for non-eksclt created clusters.

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [X] Manually tested
- [X] Added labels for change area (e.g. `area/nodegroup`), target version (e.g. `version/0.12.0`) and kind (e.g. `kind/improvement`)
- [X] Make sure the title of the PR is a good description that can go into the release notes

### CODE REVIEW

File: pkg/ctl/upgrade/cluster.go [hunk 1]
```diff
@@ -82,31 +84,39 @@ func DoUpgradeCluster(cmd *cmdutils.Cmd) error {
...
 	}
 
+	stackManager := ctl.NewStackManager(cfg)
+	stacks, err := stackManager.DescribeStacks()
+	if err != nil {
+		return err
+	}
+
+	if manager.IsClusterStack(stacks) {
+		return upgradeEKSCTLCluster(versionUpdateRequired, cmd.Plan, cfg, currentVersion, ctl, stackManager)
```
Comment (michaelbeaumont) [inline pkg/ctl/upgrade/cluster.go:101]: Should we start a `Cluster interface` that provides `Upgrade` and have two concrete types implement it? Something we can expand as more non-eksctl created operations become possible.

Comment (aclevername) [reply pkg/ctl/upgrade/cluster.go:101]: Great idea, I'll incorporate it into the PR

File: pkg/actions/cluster/owned.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+func (c *OwnedCluster) Upgrade(dryRun bool) error {
+	currentVersion := c.ctl.ControlPlaneVersion()
+	versionUpdateRequired, err := requiresVersionUpgrade(c.cfg.Metadata, currentVersion)
+	if err != nil {
+		return err
+	}
+
+	if err := c.ctl.LoadClusterVPC(c.cfg); err != nil {
```
Comment (Callisto13) [inline pkg/actions/cluster/owned.go:34]: did i peer pressure you into doing your errors like this? 😂

File: pkg/actions/cluster/owned.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	printer := printers.NewJSONPrinter()
+	if err := printer.LogObj(logger.Debug, "cfg.json = \\\n%s\n", c.cfg); err != nil {
+		return err
+	}
+
+	if versionUpdateRequired {
+		if err := updateVersion(dryRun, c.cfg, currentVersion, c.ctl); err != nil {
+			return err
+		}
+	}
```
Comment (Callisto13) [inline pkg/actions/cluster/owned.go:47]: do we want to move this up to below line 29? (or move 29 down a bit if something is needed for `updateVersion`) looks weird to do the check so far after getting the value

File: pkg/actions/cluster/owned.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		return err
+	}
+
+	if err := c.ctl.ValidateExistingNodeGroupsForCompatibility(c.cfg, c.stackManager); err != nil {
+		logger.Critical("failed checking nodegroups", err.Error())
+	}
+
+	cmdutils.LogPlanModeWarning(dryRun && (stackUpdateRequired || versionUpdateRequired))
+	return nil
+
```
Comment (Callisto13) [inline pkg/actions/cluster/owned.go:69]:
```suggestion
```

File: pkg/actions/cluster/update.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	case api.Version1_14:
+		return api.Version1_15, nil
+	case api.Version1_15:
+		return api.Version1_16, nil
+	case api.Version1_16:
+		return api.Version1_17, nil
+	case api.Version1_17:
+		return api.Version1_18, nil
+	case api.Version1_18:
+		return api.Version1_19, nil
```
Comment (Callisto13) [inline pkg/actions/cluster/update.go:92]: A couple of questions here:
1. How come you started at 1.12? We list versions as far back as 1.10 are listed as unsupported.
1. What is the UX for people whose next version is unsupported? I can see we give them this error, but do we suggest a next course of action? I can see people thinking "oh i'll just bump n versions then" which we also don't allow.

File: pkg/actions/cluster/owned.go (hunk 3 above)
Comment (aclevername) [reply pkg/actions/cluster/owned.go:47]: it sort of feels in the right place, we error out early if we can't check if theres an update, and then afterwards we load cluster information, log what we've found and then proceed with the upgrade.

File: pkg/actions/cluster/update.go (hunk 5 above)
Comment (aclevername) [reply pkg/actions/cluster/update.go:92]: Great question, I'm not sure regarding 2). I know that we do manually add versions to eksctl when new versions of kubernetes are released. 

> I can see people thinking "oh i'll just bump n versions then" which we also don't allow.

Do you mean kubernetes versions? or CLI versions?

Comment (Callisto13) [reply pkg/actions/cluster/update.go:92]: kubernetes

Comment (aclevername) [reply pkg/actions/cluster/update.go:92]: I'm assuming the EKS API will reject it, although I haven't tried.

Comment (Callisto13) [reply pkg/actions/cluster/update.go:92]: hmm i imagine that looks pretty. meh, we can wait until somebody tries it and gets confused

File: pkg/actions/cluster/unowned.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+func NewUnownedCluster(cfg *api.ClusterConfig, ctl *eks.ClusterProvider) (*UnownedCluster, error) {
+	return &UnownedCluster{
+		cfg: cfg,
+		ctl: ctl,
+	}, nil
+}
+
+func (c *UnownedCluster) Upgrade(dryRun bool) error {
```
Comment (michaelbeaumont) [inline pkg/actions/cluster/unowned.go:22]: this function seems to be duplicated in `owned.go` except for the `LoadClusterVPC` call, can that be reduced somehow?

Comment (aclevername) [reply pkg/actions/cluster/unowned.go:20]: lmk what you think @michaelbeaumont https://github.com/weaveworks/eksctl/pull/2815/commits/a6e01da33ca345697881b96da1307b538e61b92f

**Comment by aclevername** [comment]
Going to re-test the functionality once the PR is approved as a few minor refactors have been done since the last test

**Review by michaelbeaumont — APPROVED** [review]

**Comment by aclevername** [comment]
> Going to re-test the functionality once the PR is approved before merging as a few minor refactors have been done since the last test

Tested works 👍

**Pull request merged**